torch/csrc/autograd/variable.cpp

```diff
@@ -77,23 +77,30 @@ Scalar Variable::Impl::localScalar() {
 }
 
 std::shared_ptr<Function> Variable::Impl::get_grad_accumulator() {
-  if (grad_fn_) {
-    throw std::logic_error(
-        "get_grad_accumulator() should be only called on leaf Variables");
-  }
-  if (!requires_grad_) {
-    return nullptr;
-  }
+  // Short-circuits if accumulator requested for non-leaf
+  if (grad_fn_) throw std::logic_error(
+    "get_grad_accumulator() should be only called on leaf Variables");
 
-  std::lock_guard<std::mutex> lock(mutex_);
+  // Short-cicuits if no gradient is required
+  if (!requires_grad_) return nullptr;
 
-  auto result = grad_accumulator_.lock();
-  if (result)
-    return result;
+  // Checks for up-to-date accumulator
+  // Note: the accumulator's metadata may be outdated after a call to
+  // set_data, for example
+  std::lock_guard<std::mutex> lock(mutex_);
+  auto cur_accumulator = grad_accumulator_.lock();
+  const auto device = data_.is_cuda() ? data_.get_device() : -1;
+  if (cur_accumulator && cur_accumulator->input_metadata(0).matches(type(), device)) { 
```

Comment:
Can't we simply invalidate the accumulator field when `set_data` is called? This would let us avoid having to check this every time a leaf Variable is used.

Reply:
Actually I now see that what I proposed is exactly what it used to be. Why did that need to change? I think the older way is nicer

Reply:
You're right that moving the check to set_data is faster, and I did consider it. I was trying to play it safe and not make assumptions about the situations that the data_ might change previously, but if set_data() is the only way to change it then we're good. I'll get these moved.

Reply:
Yeah, it should be the only place where it changes. Can you please revert this part of the diff?

Reply:
Yep; I'll move both checks there.